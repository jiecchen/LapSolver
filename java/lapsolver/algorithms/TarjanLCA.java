/**
 * @file TarjanLCA.java
 * @author Cyril Zhang <cyril.zhang@yale.edu>
 * @date Tue Jun 3 2014
 *
 * An algorithm for computing offline LCA queries in a tree.
 * Complexity: amortized O( q alpha(n) )
 * Reference: http://en.wikipedia.org/wiki/Tarjan's_off-line_lowest_common_ancestors_algorithm
 */

package lapsolver.algorithms;

import lapsolver.Tree;

public class TarjanLCA {
    private Tree tree;

    // state variables for algorithm
    private UnionFind unionFind;
    private boolean[] black;
    private int[] ancestor;

    // store queries as a subgraph
    private int nq;
    private int[][] queries, queryIndices;
    private int[] answer;

    // initialize state
    public TarjanLCA (Tree tree) {
        this.tree = tree;

        // set up algorithm state
        unionFind = new UnionFind(tree.nv);
        black = new boolean[tree.nv];
        ancestor = new int[tree.nv];

        // set up adjacency lists for queries
        queries = new int[tree.nv][];
        queryIndices = new int[tree.nv][];
    }

    // preprocess, solve, return answers
    public int[] solve (int[] a, int[] b) {
        if (a.length != b.length) {
            throw new Error("LCA query arrays should have same length");
        }
        nq = a.length;
        answer = new int[nq];

        // one pass to compute degrees
        int[] queryDegree = new int[tree.nv];
        for (int i = 0; i < nq; i++) {
            queryDegree[a[i]]++;
            queryDegree[b[i]]++;
        }

        // init query array sizes
        for (int i = 0; i < tree.nv; i++) {
            queries[i] = new int[queryDegree[i]];
            queryIndices[i] = new int[queryDegree[i]];
        }

        // populate queries
        for (int i = 0; i < nq; i++) {
            queries[ a[i] ][ --queryDegree[a[i]] ] = b[i];
            queries[ b[i] ][ --queryDegree[b[i]] ] = a[i];

            queryIndices[ a[i] ][ queryDegree[a[i]] ] = i;
            queryIndices[ b[i] ][ queryDegree[b[i]] ] = i;
        }

        dfs(tree.root);

        return answer;
    }

    // main DFS procedure (from Wikipedia article)
    private void dfs(int u) {
        ancestor[u] = u;
        for (int v : tree.nodes[u].children) {
            dfs(v);
            unionFind.union(u, v);
            ancestor[unionFind.find(u)] = u;
        }
        black[u] = true;

        for (int i = 0; i < queries[u].length; i++) {
            int v = queries[u][i];
            if (black[v]) {
                answer[ queryIndices[u][i] ] = ancestor[unionFind.find(v)];
            }
        }
    }

}
