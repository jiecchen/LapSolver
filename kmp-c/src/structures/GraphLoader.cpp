#include "GraphLoader.h"
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

Graph GraphLoader::fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w)
{
    return Graph(EdgeList(u, v, w));
}

Graph GraphLoader::fromFile(const string &filename)
{
    aligned_vector<int> u = aligned_vector<int>();
    aligned_vector<int> v = aligned_vector<int>();
    aligned_vector<double> w = aligned_vector<double>();

    string line;
    ifstream ijvFile(filename);
    if (!ijvFile)
        throw -1;

    while (getline(ijvFile, line))
    {
        if (line.empty()) break;
        istringstream iss(line);
        int a, b;
        double c;
        if (!(iss >> a >> b >> c))
            throw u.size();
        u.push_back(a);
        v.push_back(b);
        w.push_back(c);
    }

    return fromArrays(u, v, w);
}