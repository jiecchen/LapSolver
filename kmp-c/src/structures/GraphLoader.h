#pragma once

#include <string>
#include <iostream>
#include "structures/Graph.h"
#include "util/aligned.h"

class GraphLoader
{
public:
	
	static Graph fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w);
	static Graph fromFile(const std::string &filename);
    static Graph fromStream(std::istream &inStream);
    static Graph fromStdin();

};