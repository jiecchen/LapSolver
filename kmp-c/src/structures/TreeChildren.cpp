#include "TreeChildren.h"

TreeChildren::TreeChildren(int nv, const int* parent) {
    // offset = cumulative sums of child counts
    offset = new int[nv+1];
    offset[0:nv+1] = 0;

    for (int i = 0; i < nv; i++) {
        if(parent[i] != i) offset[parent[i]+1]++;
    }
    for (int i = 1; i <= nv; i++) {
        offset[i] += offset[i-1];
    }

    // fill child array with these offsets
    child = new int[nv];
    int* childIndex = new int[nv];
    childIndex[0:nv] = 0;

    for (int i = 0; i < nv; i++) {
        int p = parent[i];
        if (p != i) {
            child[offset[p] + childIndex[p]++] = i;
        }
    }

    delete[] childIndex;
}

TreeChildren::~TreeChildren() {
    delete[] child;
    delete[] offset;
}