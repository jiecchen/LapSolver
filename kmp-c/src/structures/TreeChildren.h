// TreeChildren: a structure to compute and store child arrays from a parent array.
// child[offset[v], offset[v]+1, ..., offset[v+1]-1] are the children of vertex v.

#pragma once

struct TreeChildren {
    TreeChildren(int nv, const int* parent);
    ~TreeChildren();

    int* child;
    int* offset;
};
