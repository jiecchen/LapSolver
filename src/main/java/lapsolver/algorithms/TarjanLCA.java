/**
 * @file TarjanLCA.java
 * @author Cyril Zhang <cyril.zhang@yale.edu>
 * @date Tue Jun 3 2014
 *
 * An algorithm for computing offline LCA queries in a tree.
 * Complexity: amortized O( (numRemoved + q) alpha(numRemoved) )
 * Reference: http://en.wikipedia.org/wiki/Tarjan's_off-line_lowest_common_ancestors_algorithm
 */

package lapsolver.algorithms;

import lapsolver.Tree;
import lapsolver.util.TreeUtils;

public class TarjanLCA {
    private final Tree tree;

    // state variables for algorithm
    private final UnionFind unionFind;
    private final boolean[] black;
    private final int[] ancestor;

    private final int[][] queries;
    private final int[][] queryIndices;

    /**
     * Initialize the scratch space needed for the algorithm.
     * @param tree The input tree (weights are conductances or resistances).
     */
    public TarjanLCA (Tree tree) {
        this.tree = tree;

        // set up algorithm state
        unionFind = new UnionFind(tree.nv);
        black = new boolean[tree.nv];
        ancestor = new int[tree.nv];

        // set up adjacency lists for queries
        queries = new int[tree.nv][];
        queryIndices = new int[tree.nv][];
    }

    // preprocess, solve, return answers
    public int[] solve (int[] a, int[] b) {
        if (a.length != b.length) {
            throw new Error("LCA query arrays should have same length");
        }
        int nq = a.length;
        int[] answer = new int[nq];

        // one pass to compute degrees
        int[] queryDegree = new int[tree.nv];
        for (int i = 0; i < nq; i++) {
            queryDegree[a[i]]++;
            queryDegree[b[i]]++;
        }

        // init query array sizes
        for (int i = 0; i < tree.nv; i++) {
            queries[i] = new int[queryDegree[i]];
            queryIndices[i] = new int[queryDegree[i]];
        }

        // populate queries
        for (int i = 0; i < nq; i++) {
            queries[ a[i] ][ --queryDegree[a[i]] ] = b[i];
            queries[ b[i] ][ --queryDegree[b[i]] ] = a[i];

            queryIndices[ a[i] ][ queryDegree[a[i]] ] = i;
            queryIndices[ b[i] ][ queryDegree[b[i]] ] = i;
        }

        // do DFS
        for (int i = 0; i < tree.nv; i++) {
            ancestor[i] = i;
        }

        int[] order = TreeUtils.dfsOrder(tree);
        int[] childrenVisited = new int[tree.nv];

        for (int i = tree.nv-1; i >= 0; i--) {
            int v = order[i];
            int parent = tree.parent[v];

            if (v != tree.root) {
                childrenVisited[parent]++;
            }

            if (childrenVisited[v] == tree.children[v].length) {
                black[v] = true;

                for (int j = 0; j < queries[v].length; j++) {
                    int child = queries[v][j];
                    if (black[child]) {
                        answer[ queryIndices[v][j] ] = ancestor[unionFind.find(child)];
                    }
                }

                unionFind.union(parent, v);
                ancestor[unionFind.find(parent)] = parent;
            }
        }

        return answer;
    }

}
