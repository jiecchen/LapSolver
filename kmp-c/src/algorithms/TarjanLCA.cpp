#include <iostream>

#include "TarjanLCA.h"
#include "UnionFind.h"
#include "structures/Graph.h"
#include "structures/TreeChildren.h"

using namespace std;

TarjanLCA::TarjanLCA (const Graph& g, int tRoot, const int* tParent, const TreeChildren& tChildren) {
    // get dfs order
    int n = g.nv();
    int* dfsStack = new int[n]; dfsStack[0] = tRoot;
    int dfsStackPos = 1;
    int* dfsOrder = new int[n];
    int dfsOrderPos = 0;

    while (dfsStackPos > 0) {
        int curNode = dfsStack[--dfsStackPos];
        dfsOrder[dfsOrderPos++] = curNode;
        for (int i = tChildren.offset[curNode]; i < tChildren.offset[curNode+1]; i++) {
            dfsStack[dfsStackPos++] = tChildren.child[i];
        }
    }

    delete[] dfsStack;

    // run LCA
    int* ancestor = new int[n];
    bool* black = new bool[n];
    int* childrenLeft = new int[n];

    UnionFind components(n);

    int edgePos = 0;
    ne = g.ne() - n + 1;
    u = new int[ne];
    v = new int[ne];
    lca = new int[ne];

    black[0:n] = false; 
    childrenLeft[0:n] = tChildren.offset[1:n] - tChildren.offset[0:n];
    for (int i = 0; i < n; i++) ancestor[i] = i;

    for (int i = n-1; i >= 0; i--) {
        int curNode = dfsOrder[i];
        int parent = tParent[curNode];

        if (curNode != tRoot) {
            childrenLeft[parent]--;
        }

        // done with all children
        if (!childrenLeft[curNode]) {
            black[curNode] = true;

            for (int i = 0; i < g.degree(curNode); i++) {
                int queryNode = g.neighbor(curNode, i);
                if (black[queryNode] && tParent[queryNode] != curNode && tParent[curNode] != queryNode) {
                    u[edgePos] = curNode;
                    v[edgePos] = queryNode;
                    lca[edgePos] = ancestor[components.find_set(queryNode)];
                    edgePos++;
                }
            }

            components.link(parent, curNode);
            ancestor[components.find_set(parent)] = parent;
        }

        // push children
        for (int i = tChildren.offset[curNode]; i < tChildren.offset[curNode+1]; i++) {
            dfsStack[dfsStackPos++] = tChildren.child[i];
        }
    }

    delete[] ancestor;
    delete[] black;
    delete[] childrenLeft;
    delete[] dfsOrder;
}

TarjanLCA::~TarjanLCA () {
    delete[] u;
    delete[] v;
    delete[] lca;
}
