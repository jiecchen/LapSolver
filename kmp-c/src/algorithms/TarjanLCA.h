#pragma once

#include "structures/Graph.h"
#include "structures/TreeChildren.h"

struct TarjanLCA {
    TarjanLCA (const Graph& g, int tRoot, const int* tParent, const TreeChildren& tChildren);
    ~TarjanLCA ();

    int ne;
    int* u;
    int* v;
    int* lca;
};