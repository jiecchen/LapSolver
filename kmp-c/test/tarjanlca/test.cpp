// Test for Tarjan LCA

#include <cstdlib>
#include <iostream>
#include <string>
#include <algorithm>
#include <algorithms/TarjanLCA.h>
#include <structures/TreeChildren.h>
#include <structures/GraphLoader.h>

using namespace std;

int main(int argc, char *argv[])
{
    atexit(mkl_free_buffers);

    // read graph
    Graph g = GraphLoader::fromStdin();

    // read spanning tree
    int n = g.nv();
    vector<int> parent(n);
    int root;
    for (int i = 0; i < n; i++) {
        cin >> parent[i];
        if (parent[i] == i) root = i;
    }

    // get children
    TreeChildren tChildren(n, parent.data());

    // run lca
    TarjanLCA lca(g, root, parent.data(), tChildren);

    // canonize lca results
    for (int i = 0; i < lca.ne; i++) {
        if (lca.u[i] > lca.v[i]) swap(lca.u[i], lca.v[i]);
    }

    vector<int> ord(lca.ne);
    for (int i = 0; i < lca.ne; i++) ord[i] = i;
    sort(ord.begin(), ord.end(), [&] (int a, int b) {
        if (lca.u[a] != lca.u[b]) return lca.u[a] < lca.u[b];
        return lca.v[a] < lca.v[b];
    });

    // read off sorted lca results
    for (int i = 0; i < lca.ne; i++) {
        printf("lca(%d, %d) = %d\n", lca.u[ord[i]], lca.v[ord[i]], lca.lca[ord[i]]);
    }

    return 0;
}
