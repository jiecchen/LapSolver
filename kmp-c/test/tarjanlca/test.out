lca(0, 3) = 0
lca(0, 4) = 0
lca(0, 5) = 0
lca(0, 6) = 0
lca(1, 2) = 0
lca(1, 5) = 0
lca(1, 6) = 0
lca(2, 3) = 0
lca(2, 4) = 0
lca(3, 4) = 1
lca(3, 5) = 0
lca(3, 6) = 0
lca(4, 5) = 0
lca(4, 6) = 0
lca(5, 6) = 2
